In [246]:
import pandas as pd
import os

df = pd.read_csv("insurance_claims.csv")

df

months_as_customer  age  policy_number policy_bind_date policy_state  \
0                   328   48         521585       2014-10-17           OH   
1                   228   42         342868       2006-06-27           IN   
2                   134   29         687698       2000-09-06           OH   
3                   256   41         227811       1990-05-25           IL   
4                   228   44         367455       2014-06-06           IL   
..                  ...  ...            ...              ...          ...   
995                   3   38         941851       1991-07-16           OH   
996                 285   41         186934       2014-01-05           IL   
997                 130   34         918516       2003-02-17           OH   
998                 458   62         533940       2011-11-18           IL   
999                 456   60         556080       1996-11-11           OH   

    policy_csl  policy_deductable  policy_annual_premium  umbrella_limit  \
0      250/500               1000                1406.91               0   
1      250/500               2000                1197.22         5000000   
2      100/300               2000                1413.14         5000000   
3      250/500               2000                1415.74         6000000   
4     500/1000               1000                1583.91         6000000   
..         ...                ...                    ...             ...   
995   500/1000               1000                1310.80               0   
996    100/300               1000                1436.79               0   
997    250/500                500                1383.49         3000000   
998   500/1000               2000                1356.92         5000000   
999    250/500               1000                 766.19               0   

     insured_zip  ... police_report_available total_claim_amount injury_claim  \
0         466132  ...                     YES              71610         6510   
1         468176  ...                       ?               5070          780   
2         430632  ...                      NO              34650         7700   
3         608117  ...                      NO              63400         6340   
4         610706  ...                      NO               6500         1300   
..           ...  ...                     ...                ...          ...   
995       431289  ...                       ?              87200        17440   
996       608177  ...                       ?             108480        18080   
997       442797  ...                     YES              67500         7500   
998       441714  ...                     YES              46980         5220   
999       612260  ...                       ?               5060          460   

    property_claim vehicle_claim   auto_make  auto_model auto_year  \
0            13020         52080        Saab         92x      2004   
1              780          3510    Mercedes        E400      2007   
2             3850         23100       Dodge         RAM      2007   
3             6340         50720   Chevrolet       Tahoe      2014   
4              650          4550      Accura         RSX      2009   
..             ...           ...         ...         ...       ...   
995           8720         61040       Honda      Accord      2006   
996          18080         72320  Volkswagen      Passat      2015   
997           7500         52500      Suburu     Impreza      1996   
998           5220         36540        Audi          A5      1998   
999            920          3680    Mercedes        E400      2007   

    fraud_reported _c39  
0                Y  NaN  
1                Y  NaN  
2                N  NaN  
3                Y  NaN  
4                N  NaN  
..             ...  ...  
995              N  NaN  
996              N  NaN  
997              N  NaN  
998              N  NaN  
999              N  NaN  

[1000 rows x 40 columns]

## Esto lo saqué de R

> skim(df)
── Data Summary ────────────────────────
                           Values
Name                       df    
Number of rows             1000  
Number of columns          40    
_______________________          
Column type frequency:           
  character                21    
  logical                  1     
  numeric                  18    
________________________         
Group variables            None  

── Variable type: character ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   skim_variable           n_missing complete_rate min max empty n_unique whitespace
 1 policy_bind_date                0             1  10  10     0      951          0
 2 policy_state                    0             1   2   2     0        3          0
 3 policy_csl                      0             1   7   8     0        3          0
 4 insured_sex                     0             1   4   6     0        2          0
 5 insured_education_level         0             1   2  11     0        7          0
 6 insured_occupation              0             1   5  17     0       14          0
 7 insured_hobbies                 0             1   4  14     0       20          0
 8 insured_relationship            0             1   4  14     0        6          0
 9 incident_date                   0             1  10  10     0       60          0
10 incident_type                   0             1  10  24     0        4          0
11 collision_type                  0             1   1  15     0        4          0
12 incident_severity               0             1  10  14     0        4          0
13 authorities_contacted           0             1   4   9     0        5          0
14 incident_state                  0             1   2   2     0        7          0
15 incident_city                   0             1   8  11     0        7          0
16 incident_location               0             1  11  23     0     1000          0
17 property_damage                 0             1   1   3     0        3          0
18 police_report_available         0             1   1   3     0        3          0
19 auto_make                       0             1   3  10     0       14          0
20 auto_model                      0             1   2  14     0       39          0
21 fraud_reported                  0             1   1   1     0        2          0

── Variable type: logical ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  skim_variable n_missing complete_rate mean count
1 X_c39              1000             0  NaN ": " 

── Variable type: numeric ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   skim_variable               n_missing complete_rate        mean          sd        p0     p25     p50     p75      p100 hist 
 1 months_as_customer                  0             1     204.        115.           0     116.    200.    276.      479  ▅▇▇▃▂
 2 age                                 0             1      38.9         9.14        19      32      38      44        64  ▃▇▇▃▂
 3 policy_number                       0             1  546239.     257063.      100804  335980. 533135  759100.   999435  ▇▇▇▇▇
 4 policy_deductable                   0             1    1136         612.         500     500    1000    2000      2000  ▇▇▁▁▇
 5 policy_annual_premium               0             1    1256.        244.         433.   1090.   1257.   1416.     2048. ▁▃▇▃▁
 6 umbrella_limit                      0             1 1101000     2297407.    -1000000       0       0       0  10000000  ▇▁▁▁▁
 7 insured_zip                         0             1  501214.      71702.      430104  448404. 466446. 603251    620962  ▇▃▁▁▅
 8 capital.gains                       0             1   25126.      27872.           0       0       0   51025    100500  ▇▂▃▂▁
 9 capital.loss                        0             1  -26794.      28104.     -111100  -51500  -23250       0         0  ▁▁▅▂▇
10 incident_hour_of_the_day            0             1      11.6         6.95         0       6      12      17        23  ▇▇▆▇▇
11 number_of_vehicles_involved         0             1       1.84        1.02         1       1       1       3         4  ▇▁▁▅▁
12 bodily_injuries                     0             1       0.992       0.820        0       0       1       2         2  ▇▁▇▁▇
13 witnesses                           0             1       1.49        1.11         0       1       1       2         3  ▇▇▁▇▇
14 total_claim_amount                  0             1   52762.      26402.         100   41812.  58055   70592.   114920  ▃▂▇▅▁
15 injury_claim                        0             1    7433.       4881.           0    4295    6775   11305     21450  ▅▇▅▃▁
16 property_claim                      0             1    7400.       4825.           0    4445    6750   10885     23670  ▅▇▅▂▁
17 vehicle_claim                       0             1   37929.      18886.          70   30292.  42100   50822.    79560  ▃▂▇▆▁
18 auto_year                           0             1    2005.          6.02      1995    2000    2005    2010      2015  ▇▆▇▇▆
> 

In [247]:
# colnames del df
df.columns

Index(['months_as_customer', 'age', 'policy_number', 'policy_bind_date',
       'policy_state', 'policy_csl', 'policy_deductable',
       'policy_annual_premium', 'umbrella_limit', 'insured_zip', 'insured_sex',
       'insured_education_level', 'insured_occupation', 'insured_hobbies',
       'insured_relationship', 'capital-gains', 'capital-loss',
       'incident_date', 'incident_type', 'collision_type', 'incident_severity',
       'authorities_contacted', 'incident_state', 'incident_city',
       'incident_location', 'incident_hour_of_the_day',
       'number_of_vehicles_involved', 'property_damage', 'bodily_injuries',
       'witnesses', 'police_report_available', 'total_claim_amount',
       'injury_claim', 'property_claim', 'vehicle_claim', 'auto_make',
       'auto_model', 'auto_year', 'fraud_reported', '_c39'],
      dtype='object')

In [248]:
print("son ", len(df.columns), " columnas")

son  40  columnas


In [249]:
# primeros 7 uniques values de cada columna
for col in df.columns:
    print(col, df[col].unique()[:7])

months_as_customer [328 228 134 256 137 165  27]
age [48 42 29 41 44 39 34]
policy_number [521585 342868 687698 227811 367455 104594 413978]
policy_bind_date ['2014-10-17' '2006-06-27' '2000-09-06' '1990-05-25' '2014-06-06'
 '2006-10-12' '2000-06-04']
policy_state ['OH' 'IN' 'IL']
policy_csl ['250/500' '100/300' '500/1000']
policy_deductable [1000 2000  500]
policy_annual_premium [1406.91 1197.22 1413.14 1415.74 1583.91 1351.1  1333.35]
umbrella_limit [      0 5000000 6000000 4000000 3000000 8000000 7000000]
insured_zip [466132 468176 430632 608117 610706 478456 441716]
insured_sex ['MALE' 'FEMALE']
insured_education_level ['MD' 'PhD' 'Associate' 'Masters' 'High School' 'College' 'JD']
insured_occupation ['craft-repair' 'machine-op-inspct' 'sales' 'armed-forces' 'tech-support'
 'prof-specialty' 'other-service']
insured_hobbies ['sleeping' 'reading' 'board-games' 'bungie-jumping' 'base-jumping' 'golf'
 'camping']
insured_relationship ['husband' 'other-relative' 'own-child' 'unmarried' '

In [250]:
# ver tipo de datos de las primeras 
df.dtypes[:20]

months_as_customer           int64
age                          int64
policy_number                int64
policy_bind_date            object
policy_state                object
policy_csl                  object
policy_deductable            int64
policy_annual_premium      float64
umbrella_limit               int64
insured_zip                  int64
insured_sex                 object
insured_education_level     object
insured_occupation          object
insured_hobbies             object
insured_relationship        object
capital-gains                int64
capital-loss                 int64
incident_date               object
incident_type               object
collision_type              object
dtype: object

In [251]:
# Hacer policy_bind_date a datetime
df["policy_bind_date"] = pd.to_datetime(df["policy_bind_date"])
df["incident_date"] = pd.to_datetime(df["incident_date"])
print("policy_bind_date: ", df["policy_bind_date"].dtype)
print("incident_date: ", df["incident_date"].dtype)

policy_bind_date:  datetime64[ns]
incident_date:  datetime64[ns]


In [252]:
df.dtypes[20:40]

incident_severity               object
authorities_contacted           object
incident_state                  object
incident_city                   object
incident_location               object
incident_hour_of_the_day         int64
number_of_vehicles_involved      int64
property_damage                 object
bodily_injuries                  int64
witnesses                        int64
police_report_available         object
total_claim_amount               int64
injury_claim                     int64
property_claim                   int64
vehicle_claim                    int64
auto_make                       object
auto_model                      object
auto_year                        int64
fraud_reported                  object
_c39                           float64
dtype: object

In [253]:
# Sacar todas las columnas que tengan "?"
print(df.columns[df.isin(["?"]).any()])
print(df.columns[df.isin(["YES"]).any()])
print(df.columns[df.isin(["Y"]).any()])
print(df.columns[df.isin(["NO"]).any()])
print(df.columns[df.isin(["N"]).any()])

Index(['collision_type', 'property_damage', 'police_report_available'], dtype='object')
Index(['property_damage', 'police_report_available'], dtype='object')
Index(['fraud_reported'], dtype='object')
Index(['property_damage', 'police_report_available'], dtype='object')
Index(['fraud_reported'], dtype='object')


In [254]:
# Uniques de ['collision_type', 'property_damage', 'police_report_available']
print(df["police_report_available"].unique())
print(df["collision_type"].unique())
print(df["property_damage"].unique())
print(df["fraud_reported"].unique())

print(df["police_report_available"].value_counts())
print(df["property_damage"].value_counts())
print(df["fraud_reported"].value_counts())

['YES' '?' 'NO']
['Side Collision' '?' 'Rear Collision' 'Front Collision']
['YES' '?' 'NO']
['Y' 'N']
police_report_available
?      343
NO     343
YES    314
Name: count, dtype: int64
property_damage
?      360
NO     338
YES    302
Name: count, dtype: int64
fraud_reported
N    753
Y    247
Name: count, dtype: int64


In [255]:
# Mapear los valores "?" a "NO" en las columnas 'property_damage' y 'police_report_available'
cols_con_pregunta = ['property_damage', 'police_report_available']
for col in cols_con_pregunta:
    df[col] = df[col].replace('?', 'NO')
    
df["fraud_reported"] = df["fraud_reported"].replace('N', 'NO')
df["fraud_reported"] = df["fraud_reported"].replace('Y', 'YES')

print(df["police_report_available"].unique())
print(df["property_damage"].unique())
print(df["fraud_reported"].unique())
# count cuantas veces aparece "NO" en la columna y cuantas veces aparece "?" en la columna 'police_report_available' 
print(df["police_report_available"].value_counts())
print(df["property_damage"].value_counts())
print(df["fraud_reported"].value_counts())

['YES' 'NO']
['YES' 'NO']
['YES' 'NO']
police_report_available
NO     686
YES    314
Name: count, dtype: int64
property_damage
NO     698
YES    302
Name: count, dtype: int64
fraud_reported
NO     753
YES    247
Name: count, dtype: int64


In [256]:
df.dtypes[:20]


months_as_customer                  int64
age                                 int64
policy_number                       int64
policy_bind_date           datetime64[ns]
policy_state                       object
policy_csl                         object
policy_deductable                   int64
policy_annual_premium             float64
umbrella_limit                      int64
insured_zip                         int64
insured_sex                        object
insured_education_level            object
insured_occupation                 object
insured_hobbies                    object
insured_relationship               object
capital-gains                       int64
capital-loss                        int64
incident_date              datetime64[ns]
incident_type                      object
collision_type                     object
dtype: object

In [257]:
df.dtypes[20:40]

incident_severity               object
authorities_contacted           object
incident_state                  object
incident_city                   object
incident_location               object
incident_hour_of_the_day         int64
number_of_vehicles_involved      int64
property_damage                 object
bodily_injuries                  int64
witnesses                        int64
police_report_available         object
total_claim_amount               int64
injury_claim                     int64
property_claim                   int64
vehicle_claim                    int64
auto_make                       object
auto_model                      object
auto_year                        int64
fraud_reported                  object
_c39                           float64
dtype: object

In [258]:
df.drop(columns=["_c39"], inplace=True)
len(df.columns)

39

In [259]:
# Save the data to a csv file
df.to_csv("insurance_claims_clean.csv", index=False)
